sdks/go/pkg/beam/core/runtime/harness/statemgr.go

```diff
@@ -307,21 +345,47 @@ func (c *StateChannel) read(ctx context.Context) {
 }
 
 func (c *StateChannel) write(ctx context.Context) {
-	for req := range c.requests {
-		err := c.client.Send(req)
-		if err == nil {
-			continue // ok
+	var err error
+	var id string
+	for {
+		var req *pb.StateRequest
+		select {
```

Comment:
Good question.

Channel Select picks "ready" cases arbitrarily. Otherwise it's blocking until the first thing triggers.
Assuming both are ready, that means the client has experienced an error and the context has been cancelled. GRPC clients would send an error in this case. Unfortunately, since this depends on GRPC behavior, it's hard to test and verify.
The only way the context is cancelled is via the terminate method, which means requests and responses are nilled. This happens in a locked state, so that operation is safe. There's arguably a race on nilling c.requests, but lets worry about that in a minute.
 
However, assuming the client returns an error, then the go-routine would close, and then get a subsequent error from the Recv calls (which will always repeat their errors for sure), try to get the  responses back from the nil map (it's ok to "comma ok" look up from a nil map), and then not send anything.

If the context cancellation resolves first, then the goroutine exits as expected.

For the contentious case: If the write loop is waiting on the channel, I don't think there are issues, since it's already got a reference to the request channel.  If it happen before or after the nilling of requests, it's fine. If it happens as a race that's more problematic, but that process was going to have problems anyway. The worker would panic, and then would get restarted, which is fine.



1. waiting on a nil channel is fine, and doesn't cause problems (other than the goroutine deadlocking) https://play.golang.org/p/7NQh8StcNh-
It will block forever, which is what we want. When the context is canceled the done channel is closed and unblocks the code as intended. 
2.

Reply:
Ok, so if I understand correctly that race condition situation would only be happening when the worker is being terminated, so it's no big deal if it panics. In the other case where c.DoneCh is still called but no race condition happens, the rest of this function still executes fine (not that it makes much of a difference, since the worker is being terminated). Sounds fine to me.